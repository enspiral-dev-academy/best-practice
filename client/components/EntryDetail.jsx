import React, { useEffect } from "react";

import { wrappedWith, entriesContext, entryContext } from "../wrappers";

export function EntryDetail({ entries, entry, retrieveEntryById, match }) {
  const id = Number(match?.params?.id) || null;

  const entryFromState =
    entry?.id === id ? entry : entries.find((entry) => entry.id === id);

  useEffect(() => {
    if (!entryFromState && id) {
      retrieveEntryById(id);
    }
  }, []);

  const displayEntry = entryFromState || {};

  return (
    <div data-testid="entry">
      <h2>{displayEntry.name}</h2>
      <a href={displayEntry.link}>{displayEntry.link}</a>
      <p>{displayEntry.description}</p>
    </div>
  );
}

export default wrappedWith(entriesContext, entryContext)(EntryDetail);
