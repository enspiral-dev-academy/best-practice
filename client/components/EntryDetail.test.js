import React from "react";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";

import { EntryDetail } from "./EntryDetail";
import mockEntries from "../testing/mockEntries";

test("renders empty tags if entry not available from either context", () => {
  const entries = [];
  const entry = {};
  const { asFragment } = render(
    <EntryDetail entry={entry} entries={entries} />
  );
  expect(asFragment()).toMatchSnapshot();
});

test("renders empty tags when entry undefined and entries empty", () => {
  const entries = [];
  const { asFragment } = render(<EntryDetail entries={entries} />);
  expect(asFragment()).toMatchSnapshot();
});

test("renders empty tags if entry id doesn't match params id", () => {
  const entries = [];
  const entry = mockEntries[0];
  const match = { params: { id: 2 } };
  const retrieveEntryById = () => {};
  const { asFragment } = render(
    <EntryDetail
      match={match}
      entries={entries}
      entry={entry}
      retrieveEntryById={retrieveEntryById}
    />
  );
  expect(asFragment()).toMatchSnapshot();
});

test("renders empty tags if entry undefined and no entries found with matching id", () => {
  const entries = [mockEntries[1], mockEntries[2]];
  const match = { params: { id: 1 } };
  const retrieveEntryById = () => {};
  const { asFragment } = render(
    <EntryDetail
      match={match}
      entries={entries}
      retrieveEntryById={retrieveEntryById}
    />
  );
  expect(asFragment()).toMatchSnapshot();
});

test("renders entry when entry with matching id is provided", () => {
  const entry = mockEntries[0];
  const match = { params: { id: 1 } };
  const { asFragment } = render(<EntryDetail match={match} entry={entry} />);
  expect(asFragment()).toMatchSnapshot();
});

test("renders entry from entries state when found with matching id", () => {
  const entries = mockEntries;
  const match = { params: { id: 1 } };
  const { asFragment } = render(
    <EntryDetail match={match} entries={entries} />
  );
  expect(asFragment()).toMatchSnapshot();
});

test("renders entry with entry state over entries state", () => {
  const entries = mockEntries;
  const entry = {
    id: 1,
    name: "test entry",
    link: "https://testlink.com",
    description: "test description",
  };
  const match = { params: { id: 1 } };
  const { asFragment } = render(
    <EntryDetail match={match} entries={entries} entry={entry} />
  );
  expect(asFragment()).toMatchSnapshot();
});

test("retrieveEntryById is called if entry not found and id exists", () => {
  const entries = [];
  const entry = {};
  const retrieveEntryById = jest.fn();
  const match = { params: { id: 1 } };
  render(
    <EntryDetail
      entries={entries}
      entry={entry}
      retrieveEntryById={retrieveEntryById}
      match={match}
    />
  );
  expect(retrieveEntryById).toHaveBeenCalled();
});

test("retrieveEntryById is not called if id is null", () => {
  const entries = [];
  const entry = {};
  const retrieveEntryById = jest.fn();
  render(
    <EntryDetail
      entries={entries}
      entry={entry}
      retrieveEntryById={retrieveEntryById}
    />
  );
  expect(retrieveEntryById).not.toHaveBeenCalled();
});

test("retrieveEntryById is not called if entry is found", () => {
  const entry = mockEntries[0];
  const retrieveEntryById = jest.fn();
  const match = { params: { id: 1 } };
  render(
    <EntryDetail
      entry={entry}
      retrieveEntryById={retrieveEntryById}
      match={match}
    />
  );
  expect(retrieveEntryById).not.toHaveBeenCalled();
});
