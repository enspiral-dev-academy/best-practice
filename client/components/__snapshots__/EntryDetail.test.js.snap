// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty tags if entry id doesn't match params id 1`] = `
<DocumentFragment>
  <div
    data-testid="entry"
  >
    <h2 />
    <a />
    <p />
  </div>
</DocumentFragment>
`;

exports[`renders empty tags if entry not available from either context 1`] = `
<DocumentFragment>
  <div
    data-testid="entry"
  >
    <h2 />
    <a />
    <p />
  </div>
</DocumentFragment>
`;

exports[`renders empty tags if entry undefined and no entries found with matching id 1`] = `
<DocumentFragment>
  <div
    data-testid="entry"
  >
    <h2 />
    <a />
    <p />
  </div>
</DocumentFragment>
`;

exports[`renders empty tags when entry undefined and entries empty 1`] = `
<DocumentFragment>
  <div
    data-testid="entry"
  >
    <h2 />
    <a />
    <p />
  </div>
</DocumentFragment>
`;

exports[`renders entry from entries state when found with matching id 1`] = `
<DocumentFragment>
  <div
    data-testid="entry"
  >
    <h2>
      mocked entry 1
    </h2>
    <a
      href="https://mocked.link.com/1"
    >
      https://mocked.link.com/1
    </a>
    <p>
      mocked description 1
    </p>
  </div>
</DocumentFragment>
`;

exports[`renders entry when entry with matching id is provided 1`] = `
<DocumentFragment>
  <div
    data-testid="entry"
  >
    <h2>
      mocked entry 1
    </h2>
    <a
      href="https://mocked.link.com/1"
    >
      https://mocked.link.com/1
    </a>
    <p>
      mocked description 1
    </p>
  </div>
</DocumentFragment>
`;

exports[`renders entry with entry state over entries state 1`] = `
<DocumentFragment>
  <div
    data-testid="entry"
  >
    <h2>
      test entry
    </h2>
    <a
      href="https://testlink.com"
    >
      https://testlink.com
    </a>
    <p>
      test description
    </p>
  </div>
</DocumentFragment>
`;
